Create relevancy guide (#163)
meilisearch/documentation: guides/advanced_guides/relevancy.md

[For some in-depth explanations about the mechanism and about each default sorting rule](https://github.com/meilisearch/MeiliSearch/issues/358).

To be able to play with these rules and match them to the needs of your dataset, it is important to understand how each works and how to create new ones.

## Ranking rules

MeiliSearch has built-in ranking rules. These rules are essential to the relevance of the search engine.

Each of the rules has a role in finding the right documents for the given search query.
The order in which the rules are set in the settings affects the importance of the rule. The first rule is the most important, then the second and so on. By default, Meilisearch has these rules in a specific order, thinking it meets the most standard needs. This order can be changed in the settings to fit your needs.

Using a [bucket sort](/guides/advanced_guides/bucket_sort.md) algorithm, MeiliSearch uses rule by rule to find documents, until the desired amount of documents is obtained.

Here is the list of all the rules that are executed in this specific order by default:

#### 1. Typo

> the `_typo` rule sorts by ascending numbers of typos.

Reviewer: ```suggestion

The `_typo` rule sorts by ascending numbers of typos.
```

This means that a matched word with fewer typos is more relevant than a matched word with more typos.

#### 2. Words

> The `_words` rule sort by decreasing numbers of matched query words.

Reviewer: ```suggestion

The `_words` rule sorts by decreasing numbers of matched query words.
```

#### 3. Proximity

> The `_proximity` rule sort by increasing proximity of query words in hits.

Reviewer: ```suggestion

The `_proximity` rule sorts by increasing the proximity of query words in hits.
```

#### 4. Attribute

The `_attribute` rule sorts by ascending [attribute importance](/guides/advanced_guides/relevancy.md#attributes-importance).

#### 5. Words position

> The `_words_position` sort according to the position of query words in the attribute. The start is better than the end.

Reviewer: ```suggestion
The `_words_position` rule sorts according to the position of query words in the attribute. The start is better than the end.
```

#### 6. Exact

> The `_exact` - Sort by the similarity of the matched words with the query words. Words that are exactly the same are better than prefixes.

Reviewer: ```suggestion

The `_exact` rule sorts by the similarity of the matched words with the query words. Words that are exactly the same are better than prefixes.
```

#### Examples

:::: tabs

::: tab Typo

![Image from alias](../../public/image/vogli3.png)

### Typo

> - `vogli` : 0 typo

Reviewer: ```suggestion
- `vogli`: 0 typo
```

> - `volli` : 1 typo

Reviewer: ```suggestion
- `volli`: 1 typo
```

> The typo sorts the results by ascending typos on matched query words.

Reviewer: ```suggestion
The `typo` rule sorts the results by ascending typos on matched query words.
```

> Because `typo` is set as the first rule of our ranking rules, the number of `typos` sorts the document before

Reviewer: not sure to understand. Seems like the sentence is not over.

:::

::: tab Words
![Image from alias](../../public/image/new_house.png)

### Words

Matched words :
- **The Housemaid**: `Housemaid`, `new`, `house`, `housemaid`, `housework`: 5 words.
- **The open House**: `House`, `new`, `house`: 3 words.

> The `words` ranking rule sorts the results by descending number of matching query words.

Reviewer: ```suggestion
The `words` rule sorts the results by descending number of matching query words.
```

Reviewer: (I removed it because you don't use in the same kind of sentences you did. But grammatically and technically it's right of course)

:::

::: tab Proximity
![Image from alias](../../public/image/new_road.png)

### Proximity

> The reason why `Creature` is listed before `Mississippi Grind` is because of the `proximity` rule.

Reviewer: ```suggestion
The reason why `Creature` is listed before `Mississippi Grind` is because of the `proximity` rule.<br>
```

> The smallest **distance** between the matching words in `creature` is smaller than the smallest **distance** between the matching words in `Mississippi Grind`

Reviewer: ```suggestion
The smallest **distance** between the matching words in `creature` is smaller than the smallest **distance** between the matching words in `Mississippi Grind`.
```

> The proximity rule sorts by descending order of distance length between two matches.

Reviewer: ```suggestion
The `proximity` rule sorts by descending order of distance length between two matches.
```

The `attribute` rule sorts by ascending [attribute importance](/guides/advanced_guides/relevancy.md#attributes-importance).

:::

::: tab Words position

![Image from alias](../../public/image/belgium.png)

### Word position

`Gangsta` appears sooner than `Dunkirk` because `Belgium` appears sooner in the attribute.

The `word position` rule orders by ascending matching word's index number.

:::

::: tab Exactness
![Image from alias](../../public/image/knight.png)

**Exactness**

`Knight Moves` is displayed before `Knights of Badassdom` because `Knight` is a finished word, exactly the same as the search query. While with `Knights`, the search query is just a prefix.

:::

::::

## Order of the rules

Depending on your needs, you might want to change the order in which the rules are processed.

In this case, using the [settings route](/references/ranking_rules.md#add-or-replace-ranking-rules) of your index, you can change the ranking order of the sorting rules.

```json
[
  "_typo",
  "_attribute",
  "_proximity",
  "_words",
  "_words_position",
  "_exact"
]
```

## Adding your rules

New rules can be added to the existing list at any time and anywhere in the existing order.

> A custom rule lets you create a descending or ascending sorting rule on a given attribute.

Reviewer: ```suggestion
A custom rule lets you create a descending or ascending sorting rule on a given attribute.<br>
```

> To create a rule, you will need to communicate the attribute on which the rule is created and the order in which it will be sorted: `Asc(attribute_name)` or `Desc(attribute_name)`.

Reviewer: ```suggestion
To create a rule, you need to communicate the attribute on which the rule is created and the order in which it will be sorted: `Asc(attribute_name)` or `Desc(attribute_name)`.
```

This rule must be added to the existing list of ranking rules using the [settings route](/references/ranking_rules.md#add-or-replace-ranking-rules).

#### Example

> 

Reviewer: I would add a sentence to precise "Let's say we have a dataset of movies. The documents contain the fields `release_date` and `movie_ranking`." Not exactly this one, choose a one you like of course.
The idea is to confirm (for the user) that your custom ranking rule corresponds to the field of your dataset, and that is not "magic".

```
Desc(release_date)
```
This will create a rule that makes recent movies more relevant than older ones.

```
Asc(movie_ranking)
```
This will create a rule that makes movies with a good rank more relevant than others.

To add this newly created rule to the existing ranking rule, using the [settings route](/references/ranking_rules.md#add-or-replace-ranking-rules), you need to add the rule in the existing order array.

```json
[
  "_typo",
  "_attribute",
  "_proximity",
  "_words",
  "_words_position",
  "_exact",
  "Desc(release_date)",
  "Asc(movie_ranking)"
]
```

## Attributes importance

In a dataset, some fields are more relevant to the search than others. A `title`, for example, has a value more meaningful to a movie search than its `description` or its `director` name.

By default, MeiliSearch will determine the order of importance of the attributes based on the order in which they appear in the first document added. Then, each new attribute present in new documents will be added to the end of this ordered list.

> This means that, for this inference to work, the order of the fields in your dataset must be ordered before they are added to MeiliSearch. Or at least the first document.

Reviewer: I bugged on "for this inference to work", I mean I had to read this sentence multiple time. But maybe it's the right way to turn the sentence in English.

Author: Maybe if we remove the "This means that" And just start the sentence with "For this inference to work". What do you think?

### Changing the attributes order

Possibly, you want to change the order after the documents have been added. This is still very possible.

When a document is added to MeiliSearch, every new attribute inside will be added to two lists :
- [Searchable attributes list](/references/searchable_attributes.md): attributes in which to search for matching query words.

> - [Displayed attributes list](/references/displayed_attributes.md): attributes to send back in each document

Reviewer: ```suggestion
- [Displayed attributes list](/references/displayed_attributes.md): attributes to send back in each document.
```

> The once that concerns this section is the Searchable attributes list.

Reviewer: ```suggestion
The once that concerns this section is the searchable attributes list.
```

Reviewer: or:

> The once that concerns this section is the Searchable attributes list.

Reviewer: ```suggestion
The once that concerns this section is the `searchable attributes` list.
```

To be able to play with these rules and match them to the needs of your dataset, it is important to understand how each works and how to create new ones.

## Ranking rules

MeiliSearch has built-in ranking rules. These rules are essential to the relevance of the search engine.

Each of the rules has a role in finding the right documents for the given search query.
The order in which the rules are set in the settings affects the importance of the rule. The first rule is the most important, then the second and so on. By default, Meilisearch has these rules in a specific order, thinking it meets the most standard needs. This order can be changed in the settings to fit your needs.

Here is the list of all the rules that are executed in this specific order by default:

#### 1. Typo

The `typo` rule sorts by ascending numbers of typos.

This means that a matched word with fewer typos is more relevant than a matched word with more typos.

#### 2. Words

The `words` rule sorts by decreasing numbers of matched query words.

The more words of the search query present in the document, the more relevant it is. Having multiple times the same word does not affect the relevancy.

::: warning

> On this version: v.0.9.0 , a document that has a missing search query word will not be returned. Thus, this rule does not impact the search results. <Badge text="soon" type="warn"/>

Reviewer 2: ```suggestion
As of today, all query words must be present in the returned documents. This rule does not impact the search results yet. <Badge text="soon" type="warn"/>
```
